toolbox/deps_update/main.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"istio.io/test-infra/toolbox/util"
+)
+
+var (
+	repo       = flag.String("repo", "", "Update dependencies of only this repository")
+	owner      = flag.String("owner", "istio", "Github Owner or org.")
+	tokenFile  = flag.String("token_file", "", "File containing Github API Access Token.")
+	githubClnt *githubClient
+)
+
+const (
+	istioDepsFile = "istio.deps"
+)
+
+// Update the commit SHA reference in a given line from dependency file
+// to the latest stable version
+// Returns the updated line
+func replaceCommit(line string, dep dependency) (string, error) {
+	idx := strings.Index(line, "\"")
+	return line[:idx] + "\"" + dep.LastStableSHA + "\",", nil
+}
+
+// Generates an MD5 digest of the version set of the repo dependencies
+// useful in avoiding making duplicate branches of the same code change
+// Also updates dependency objects deserialized from istio.deps
+func fingerPrintAndUpdateDepSHA(repo string, deps *[]dependency) (string, error) {
+	digest, err := githubClnt.getHeadCommitSHA(repo, "master")
+	if err != nil {
+		return "", err
+	}
+	for i, dep := range *deps {
+		commitSHA, err := githubClnt.getHeadCommitSHA(dep.RepoName, dep.ProdBranch)
+		if err != nil {
+			return "", err
+		}
+		digest = digest + commitSHA
+		(*deps)[i].LastStableSHA = commitSHA
+	}
+	return util.GetMD5Hash(digest), nil
+}
+
+// Update the commit SHA reference in the dependency file of dep
+func updateDepFile(dep dependency) error {
+	input, err := ioutil.ReadFile(dep.File)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(string(input), "\n")
+	for i, line := range lines {
+		if strings.Contains(line, dep.Name+" = ") {
+			if lines[i], err = replaceCommit(line, dep); err != nil {
+				return err
+			}
+		}
+	}
+	output := strings.Join(lines, "\n")
+	return ioutil.WriteFile(dep.File, []byte(output), 0600)
+}
+
+// Assumes at the root of istio directory
+// Runs the updateVersion.sh script
+func updateIstioDeps() error {
+	pilotSHA, err := githubClnt.getHeadCommitSHA("pilot", "stable")
```

Comment:
There has to be some hadcoding in order to use `install/updateVersion.sh` because each dependency is passed in with different flag and hub. Such flags and hubs are only applicable to istio/istio so they should not be put into the json file. I did refactor this such that SHAs are read from deps array (read from json file) but I still have to specify the name such as "pilot" and "mixer" to properly use the flags. Let me know if you have smarter ways to do this.